airbyte-cdk/python/airbyte_cdk/sources/streams/core.py

```diff
@@ -193,25 +193,54 @@ def read(  # type: ignore  # ignoring typing for ConnectorStateManager because o
 
                     checkpoint_interval = self.state_checkpoint_interval
                     checkpoint = checkpoint_reader.get_checkpoint()
-                    if checkpoint_interval and record_counter % checkpoint_interval == 0 and checkpoint is not None:
+                    if checkpoint_interval and record_counter % checkpoint_interval == 0 and checkpoint is not None and state_manager:
                         airbyte_state_message = self._checkpoint_state(checkpoint, state_manager=state_manager)
                         yield airbyte_state_message
 
                     if internal_config.is_limit_reached(record_counter):
                         break
             self._observe_state(checkpoint_reader)
             checkpoint_state = checkpoint_reader.get_checkpoint()
-            if checkpoint_state is not None:
+            if state_manager and checkpoint_state is not None:
```

Comment:
nit: we could refactor this predicate in a method `_should_checkpoint` to ensure the logic stays aligned with line 215

Reply:
given it's simplicity of the predicate, how about we just define a variable at the beginning called `should_checkpoint` and we pass that in, subtle difference to what we have but still conveys the same idea that each condition is based on a single predicate defined at the beginning